rkt/api_service.go

```diff
@@ -0,0 +1,557 @@
+	}
+
+	// Get pod's pid if it's running.
+	pid := -1
+	if state == v1alpha.PodState_POD_STATE_RUNNING {
+		pid, err = p.getPID()
+		if err != nil {
+			log.Printf("Failed to get PID for pod %q: %v", p.uuid, err)
+			return nil, nil, err
+		}
+	}
+
+	// Get app list.
+	var apps []*v1alpha.App
+	applist, err := p.getApps()
+	if err != nil {
+		log.Printf("Failed to get app list for pod %q: %v", p.uuid, err)
+		return nil, nil, err
+	}
+
+	for _, app := range applist {
+		img := &v1alpha.Image{
+			BaseFormat: &v1alpha.ImageFormat{
+				// Only support appc image now. If it's a docker image, then it
+				// will be transformed to appc before storing in the disk store.
+				Type:    v1alpha.ImageType_IMAGE_TYPE_APPC,
+				Version: schema.AppContainerVersion.String(),
+			},
+			Id: app.Image.ID.String(),
+			// Only image format and image ID are returned in 'ListPods()'.
+		}
+
+		apps = append(apps, &v1alpha.App{
+			Name:  app.Name.String(),
+			Image: img,
+			// State and exit code are not returned in 'ListPods()'.
+		})
+	}
+
+	// Get network info.
+	var networks []*v1alpha.Network
+	for _, n := range p.nets {
+		networks = append(networks, &v1alpha.Network{
+			Name: n.NetName,
+			// There will be IPv6 support soon so distinguish between v4 and v6
+			Ipv4: n.IP.String(),
+		})
+	}
+
+	return &v1alpha.Pod{
+		Id:       p.uuid.String(),
+		Pid:      int32(pid),
+		State:    state,
+		Apps:     apps,
+		Manifest: data,
+		Networks: networks,
+	}, &manifest, nil
+}
+
+func (s *v1AlphaAPIServer) ListPods(ctx context.Context, request *v1alpha.ListPodsRequest) (*v1alpha.ListPodsResponse, error) {
+	var pods []*v1alpha.Pod
+	if err := walkPods(includeMostDirs, func(p *pod) {
+		// Get basic pod info.
+		pod, manifest, err := getBasicPod(p)
+		if err != nil { // Do not return partial pods.
+			return
+		}
+
+		// Filter the pod.
+		if !filterPod(pod, manifest, request.Filter) {
+			pods = append(pods, pod)
+		}
+	}); err != nil {
+		log.Printf("Failed to list pod: %v", err)
+		return nil, err
+	}
+	return &v1alpha.ListPodsResponse{Pods: pods}, nil
+}
+
+func (s *v1AlphaAPIServer) InspectPod(ctx context.Context, request *v1alpha.InspectPodRequest) (*v1alpha.InspectPodResponse, error) {
+	uuid, err := types.NewUUID(request.Id)
+	if err != nil {
+		log.Printf("Invalid pod id %q: %v", request.Id, err)
+		return nil, err
+	}
+
+	p, err := getPod(uuid)
+	if err != nil {
+		log.Printf("Failed to get pod %q: %v", request.Id, err)
+		return nil, err
+	}
+	defer p.Close()
+
+	// Get basic pod info.
+	pod, _, err := getBasicPod(p)
+	if err != nil {
+		return nil, err
+	}
+
+	statusDir, err := p.getStatusDir()
+	if err != nil {
+		log.Printf("Failed to get pod exit status directory: %v", err)
+		return nil, err
+	}
+
+	// Fill the extra pod info that is not available in ListPods().
+	for _, app := range pod.Apps {
+		// Fill app's image info.
+		aciInfo, err := s.store.GetACIInfoWithBlobKey(app.Image.Id)
+		if err != nil {
```

Comment:
What if the pod is still there but the image has been removed from the store? Would we have a similar problem as https://github.com/coreos/rkt/issues/1628?